### ISSUE TITLE
[Feature]: random delay before spec starts to mitigate DDoS

### ISSUE THREAD

**Issue by dimkin-eu** [microsoft/playwright#29720]
### 🚀 Feature Request

Having more tests one adds more CPU to make parallel runs to make the test suite pass faster. 
More CPU means more threads, which causes some kind of DDoS for the server ( which usually is not such beefy as prod and not mentioned to hold some load tests) because all test specs are started simultaneously. Having 16 workers sometimes makes CI to suffer :)
Is there a way to start each spec with some random delay to spread up the initial pressure of the server? 

### Motivation

more stable runs

**Comment by pavelfeldman** [comment]
Do you think running a server per worker would be a better solution? We are exploring the ideas in the area.

**Comment by dgozman** [comment]
@dimkin-eu Any thoughts on the above?

**Comment by dimkin-eu** [comment]
@dgozman, best solution would be - first tests could be delayed with some predefined pause by the test runner.
- user has 20 specs
- user has 5 workers
tests are starting and 1st test starts immediately, 2nd - after 1s, 3rd after 2s and so on till all workers are utilised.

**Comment by dgozman** [comment]
@dimkin-eu I understand your proposal. However, we still would like to know whether you have considered running a server per worker?

As a workaround, you can achieve the pause behavior with an auto-worker fixture:

```ts
import { test as base } from '@playwright/test';

export const test = base.extend<{}, { delay: void }>({
  delay: [async ({}, use, info) => {
    const seconds = info.workerIndex < info.config.workers ? info.workerIndex : 0;
    await new Promise(r => setTimeout(r, seconds * 1000));
    await use();
  }, { scope: 'worker', auto: true, timeout: 0 }],
});
```

Let me know whether that works for you.

**Comment by dimkin-eu** [comment]
@dgozman can you elaborate on `server per worker` ? we are running our tests using k8s, so a new pod with some CPUs ( depending on some calculations) is created. Is  `server per worker` some kind of sharding?

**Comment by dgozman** [comment]
@dimkin-eu Conceptually, you can have a single server for all the tests, either external or started by playwright with `webServer` option, or have multiple servers where each test talks to one server, effectively balancing the load on the server. We do not have anything built-in for this right now, so we are interested whether someone is already doing that, and/or what problems they have encountered/foresee with such an approach.

**Comment by dimkin-eu** [comment]
Our SUT is a bunch of services ( created for every test run in most MRs), so running a bunch of local ones ain't possible. The only thing we considering now - dig deeper to sharding possibilities, but this is something different.
IIRC there was something similar to our needs in  [jmeter](https://www.blazemeter.com/blog/jmeter-ramp-up-period)
```
1000 target threads with 1000 seconds ramp-up: JMeter will add one user each second
1000 target threads with 100 seconds ramp-up: JMeter will add 10 users each second
1000 target threads with 50 seconds ramp-up: JMeter will add 20 users each second
```

**Comment by dgozman** [comment]
@dimkin-eu I see, thank you for the information. Let me know whether the workaround above works for you. Meanwhile, we'll think about your usecase and similar ones, and any possible solutions we can provide.

**Comment by dgozman** [comment]
@dimkin-eu Yep, that's right. You can also re-export `expect` from the helper file, so that you write:

```ts
import { test, expect } from './my-fixtures.ts';
```

**Comment by dimkin-eu** [comment]
That is very cumbersome and unsafe ( having 200+ specs ) :(

**Comment by mstepin** [comment]
If I could add a consideration here. It would also make sense to be able to ramp up / throttle the execution of fixtures. We even limit our CI runs to just 3 workers, however as our software grows, the number of fixtures we have grow. As tests start, all fixtures that are required (auto = true) will fire up simultaneously (well almost, it obviously relies on any underlying dependancy chain that fixtures do so beautifully). This obviously strains the S.U.T in the same way that having lots of workers for lots of tests would.

**Comment by pavelfeldman** [comment]
#### Why was this issue closed?

Thank you for your involvement. This issue was closed due to limited engagement (upvotes/activity), lack of recent activity, and insufficient actionability. To maintain a manageable database, we prioritize issues based on these factors.

If you have additional information not present in this issue that you think will help prioritizing it, please open a new issue and reference this one. More support or clarity on its necessity may prompt a review. Your understanding and cooperation are appreciated.
